integration/postgres.go

```diff
@@ -0,0 +1,196 @@
+		// define the postgres image version
+		fmt.Sprintf("postgres:%s", db.dbVersion),
+		// define the arguments to docker
+		[]string{
+			fmt.Sprintf("--name=%s", containerName),
+			fmt.Sprintf("--publish=%d:5432", db.port),
+			fmt.Sprintf("--env=POSTGRES_DB=%s", db.dbName),
+			fmt.Sprintf("--env=POSTGRES_PASSWORD=%s", db.dbPassword),
+			fmt.Sprintf("--env=POSTGRES_USER=%s", db.dbUser),
+			"--detach",
+			"--rm",
+		},
+		// define the runtime arguments to postgres
+		[]string{},
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := db.checkConnection(); err != nil {
+		cleanup()
+		return nil, err
+	}
+	return cleanup, nil
+}
+
+func (db testPostgresDB) checkConnection() error {
+	ctx, cancel := context.WithTimeout(context.Background(), db.timeout)
+	defer cancel()
+	errStream := make(chan error)
+	go func() {
+		for {
+			select {
+			case <-time.After(500 * time.Millisecond):
+				driver, _ := sql.Open("postgres", db.GetDSN())
+				if err := driver.Ping(); err == nil {
+					errStream <- nil
+					return
+				}
+			case <-ctx.Done():
+				errStream <- errConnectionTimeout
+			}
+		}
+	}()
+	if err := <-errStream; err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetDSN returns the database connection string for the test Postgres database
+func (db testPostgresDB) GetDSN() string {
+	return fmt.Sprintf(
+		"host=localhost port=%d user=%s password=%s sslmode=disable dbname=%s",
+		db.port, db.dbUser, db.dbPassword, db.dbName,
+	)
+}
+
+type option func(*testPostgresDB)
+
+// WithName is used to specify the name of the test Postgres database. By default
+// the database name is "test-postgres-db"
+func WithName(name string) func(*testPostgresDB) {
+	return func(db *testPostgresDB) {
+		db.dbName = name
+	}
+}
+
+// WithPort is used to specify the port of the test Postgres database. By
+// default, the test database will find the first open port in the 35000+ range
+func WithPort(port int) func(*testPostgresDB) {
+	return func(db *testPostgresDB) {
+		db.port = port
+	}
+}
+
+// WithUser is used to specify the name of the Postgres user that owns the
+// test database
+func WithUser(user string) func(*testPostgresDB) {
+	return func(db *testPostgresDB) {
+		db.dbUser = user
+	}
+}
+
+// WithPassword is used to specify the password of the test Postgres database
+func WithPassword(password string) func(*testPostgresDB) {
+	return func(db *testPostgresDB) {
+		db.dbPassword = password
+	}
+}
+
+// WithVersion is used to specify the version of the test Postgres database. By
+// default the version is "latest"
+func WithVersion(version string) func(*testPostgresDB) {
+	return func(db *testPostgresDB) {
+		db.dbVersion = version
+	}
+}
+
+// WithMigrateFunction is used to rebuild the test Postgres database on a
+// per-test basis. Whenever the database is reset with the Reset() function, the
+// migrate function will rebuild the tables.
+func WithMigrateFunction(migrateFunction func(*sql.DB) error) func(*testPostgresDB) {
```

Comment:
I was thinking about this more, and maybe it'd be nice to have a `WithUnMigrateFunction` or something that would tear down the database as well. The default setting can be your `DROP SCHEMA` command, but a user could, for instance, pass in the `golang-migrate` `migrate.down` command into it for finer control